### PR TITLE
Fix "variable does not need to be mutable" warning

### LINKED ISSUE #534: `struct S {}`: variable does not need to be mutable
```rust
#![feature(plugin, custom_derive)]
#![plugin(serde_macros)]

#[derive(Serialize)]
struct S {}

fn main() {
    let _ = S {};
}
```

```
warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
 --> src/main.rs:4:10
  |
4 | #[derive(Serialize)]
  |          ^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #534.

cc @EliDupree

### CODE REVIEW

File: serde_codegen/src/ser.rs [hunk 1]
```diff
@@ -226,9 +227,10 @@ fn serialize_tuple_struct(
 
     let type_name = name_expr(builder, item_attrs.name());
     let len = serialize_stmts.len();
+    let let_mut = mut_if_nonzero(cx, len);
 
     quote_block!(cx, {
-        let mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
+        let $let_mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
         $serialize_stmts
```
Comment (dtolnay) [inline serde_codegen/src/ser.rs]: @nox suggests rolling this into `$serialize_stmts` and generating:

```rust
let state = try!(...);
let mut state = state;
```

where the initial call is never mut and `$serialize_stmts` includes "let mut state = state" if necessary.

Comment (dtolnay) [inline serde_codegen/src/ser.rs]: @nox I think I would rather move in the opposite direction, of making our generated code more straightforward where that can be accomplished easily. See as examples #479 and #538. There are a few reasons that is valuable:

1. The most straightforward code will give us (and others) the best experience when the generated code needs to be debugged.
2. The generated code is valuable as a template for implementing Serialize and Deserialize manually with minor modifications.
3. Convoluted generated code gives people the impression that Serde is hard or Serde codegen is magic and impossible to understand, scaring people away from implementing Serialize and Deserialize manually where necessary.

I'm not saying we need to make the generated code beautiful at the expense of maintainability of serde_codegen, but there is a middle ground and I don't think rolling `let mut state = state` into `$serialize_stmts` is the right tradeoff in this case. Thanks for the idea though!

File: serde_codegen/src/ser.rs [hunk 2]
```diff
@@ -253,8 +255,12 @@ fn serialize_struct(
     );
 
     let type_name = name_expr(builder, item_attrs.name());
-    let len = fields.iter()
+
+    let serialized_fields: Vec<_> = fields.iter()
```
Comment (oli-obk) [inline serde_codegen/src/ser.rs]: we could eliminate the allocation by folding into a tuple, not sure if that makes any difference in readability or performance, so feel free to merge as is.

Comment (nox) [inline serde_codegen/src/ser.rs]: We could also make a `Peekable` iterator and check if `fields.peek().is_none()` is true, no?

Comment (oli-obk) [inline serde_codegen/src/ser.rs]: I like it, much cleaner!

Comment (dtolnay) [inline serde_codegen/src/ser.rs]: Done. Great idea.

**Comment by oli-obk** [comment]
@homu r+

**Pull request merged**